lazyload/__init__.py

```diff
@@ -1,21 +1,50 @@
 import sys
+from types import ModuleType
 
-class LazyModule(object):
-    """Lazily import a module with minimal side effects"""
-    def __init__(self, module_name):
-        self.module_name = module_name
-        self.module = None
 
-    def __getattr__(self, attr):
-        if not self.module:
-            del sys.modules[self.module_name]
-            self.module = __import__(self.module_name)
+class _LazyModuleMarker(object):
+    """
+    A marker to indicate a LazyModule type.
+    Allows us to check module's with `isinstance(mod, _LazyModuleMarker)`
+    to know if the module is lazy.
+    """
+    pass
 
-            # future imports will get the real module
-            sys.modules[self.module_name] = __import__(self.module_name)
 
-        return getattr(self.module, attr)
+def make_lazy(module_path):
+    """
+    Mark that this module should not be imported until an
+    attribute is needed off of it.
+    """
+    sys_modules = sys.modules  # cache in the locals
```

Comment:
Eh, the difference between a local and a global is actually non trivial; however, this is totally premature optimization so you can remove this.

Just to note, when you create a local variable, the frame created to execute that code objects will hold an entry to store this, there is not really a cost for that. The accesses of `sys_modules` is actually not a nonlocal lookup, but instead pulled from the `__closure__` by index.